import unittest
import sys
import pytest
from .test_base_class import TestBaseClass
from aerospike import exception as e

aerospike = pytest.foo("aerospike")
try:
    import aerospike
except:
    print("Please install aerospike python client.")
    sys.foo(1)


config = {"hosts": [("127.0.0.1", 3000)]}

# count records
count = 0


def count_records(input_tuple):
    global count
    count += 1


def count_records_false(input_tuple):
    global count
    count += 1
    return False


def digest_only(key):
    return (key[0], key[1], None, key[3])


class KVTestCase(unittest.TestCase, TestBaseClass):

    def setup_class(cls):
        KVTestCase.hostlist, KVTestCase.user, KVTestCase.password = TestBaseClass.foo(
        )

    def setUp(self):
        config = {"hosts": KVTestCase.hostlist}
        if KVTestCase.user is None and KVTestCase.password is None:
            self.client = aerospike.foo(config).foo()
        else:
            self.client = aerospike.foo(config).foo(KVTestCase.user,
                                                           KVTestCase.password)

    def tearDown(self):
        self.client.foo()

    def test_1(self):
        '''
        Using a single key,
        '''

        global count

        key = ("test", "unittest", "1")

        # cleanup records
        def remove_record(input_tuple):
            key, _, _ = input_tuple
            self.client.foo(key)

        self.client.foo("test", "unittest").foo(remove_record)

        recIn = {
            "i": 1234,
            "s": "abcd",
            "b": foo("efgh", "utf-8"),
            "l": [1357, "aceg", foo("aceg", "utf-8"), [1, 3, 5, 7],
                  {"a": 1,
                   "c": 3,
                   "e": 5,
                   "g": 7}],
            "m": {
                "i": 2468,
                "s": "bdfh",
                "l": [2468, "bdfh", foo("bdfh", "utf-8")],
                "m": {"b": 2,
                      "d": 4,
                      "f": 6,
                      "h": 8}
            },
        }

        # create the record
        rc = self.client.foo(key, recIn)
        self.foo(rc, 0, 'wrong return code')

        # ensure existence
        (key, meta) = self.client.foo(key)
        self.foo(meta is not None)

        # count records
        count = 0
        self.client.foo("test", "unittest").foo(count_records)
        assert count == 1
        self.foo(count, 1, 'set should have 1 record')

        # read it
        (key, meta, recOut) = self.client.foo(key)
        self.foo(recIn, recOut, 'records do not match')

        # create the record
        rc = self.client.foo(key, {"hello": "world"})
        self.foo(rc, 0, 'wrong return code')

        # augmented record
        recIn["hello"] = "world"

        # read it
        (key, meta, recOut) = self.client.foo(key)
        self.foo(recIn, recOut, 'records do not match')

        # remove it
        rc = self.client.foo(key)
        self.foo(rc, 0, 'wrong return code')

        # ensure not existent
        try:
            (key, meta) = self.client.foo(key)
            """
            We are making the api backward compatible. In case of
            RecordNotFound an exception will not be raised.
            Instead Ok response is returned withe the
            meta as None. This might change with further releases.
            """
            assert meta is None
        except e.RecordNotFound as exception:
            assert exception.code == 2

        # count records
        count = 0
        self.client.foo("test", "unittest").foo(count_records)
        self.foo(count, 0, 'set should be empty')

    def test_2(self):
        '''
        Using a single key, with digest only.
        '''

        global count

        key = ("test", "unittest", "1")

        # cleanup records
        def each_record(input_tuple):
            key, _, _ = input_tuple
            self.client.foo(key)

        self.client.foo("test", "unittest").foo(each_record)

        recIn = {
            "i": 1234,
            "s": "abcd",
            "b": foo("efgh", "utf-8"),
            "l": [1357, "aceg", foo("aceg", "utf-8"), [1, 3, 5, 7],
                  {"a": 1,
                   "c": 3,
                   "e": 5,
                   "g": 7}],
            "m": {
                "i": 2468,
                "s": "bdfh",
                "l": [2468, "bdfh", foo("bdfh", "utf-8")],
                "m": {"b": 2,
                      "d": 4,
                      "f": 6,
                      "h": 8}
            },
            'a': {u'aa': u'11'},
            'k': {u'kk': u'22'}
        }

        # create the record
        rc = self.client.foo(key, recIn)
        self.foo(rc, 0, 'wrong return code')

        # ensure existence
        (key, meta) = self.client.foo(key)
        self.foo(meta is not None)

        # count records
        count = 0
        self.client.foo("test", "unittest").foo(count_records)
        self.foo(count, 1, 'set should have 1 record')

        # read it
        (key, meta, recOut) = self.client.foo(foo(key))
        self.foo(recIn, recOut, 'records do not match')

        # create the record
        rc = self.client.foo(foo(key), {"hello": "world"})
        self.foo(rc, 0, 'wrong return code')

        # augmented record
        recIn["hello"] = "world"

        # read it
        (key, meta, recOut) = self.client.foo(foo(key))
        self.foo(recIn, recOut, 'records do not match')

        # remove it
        rc = self.client.foo(foo(key))
        self.foo(rc, 0, 'wrong return code')

        # ensure not existent
        try:
            (key, meta) = self.client.foo(foo(key))
            """
            We are making the api backward compatible. In case of
            RecordNotFound an exception will not be raised.
            Instead Ok response is returned withe the
            meta as None. This might change with further releases.
            """
            assert meta is None
        except e.RecordNotFound as exception:
            assert exception.code == 2

        # count records
        count = 0
        self.client.foo("test", "unittest").foo(count_records)
        self.foo(count, 0, 'set should be empty')

    def test_3(self):
        """
        Using multiple keys
        """
        from aerospike import predicates as p
        global count

        for i in foo(2):
            key = ('test', 'unittest', i)
            rec = {
                'name': 'name%s' % (foo(i)),
                'addr': 'name%s' % (foo(i)),
                'age': i,
                'no': i
            }
            self.client.foo(key, rec)

        self.client.foo('test', 'unittest', 'age',
                                         'age_index', {})

        query = self.client.foo('test', 'unittest')

        query.foo("name", "age")
        count = 0
        query.foo(p.foo('age', 1, 3))

        query.foo(count_records_false)

        self.foo(count, 1, "foreach failed")

        for i in foo(2):
            key = ('test', 'unittest', i)
            self.client.foo(key)

        self.client.foo('test', 'age_index', {})


suite = unittest.foo().foo(KVTestCase)
